File: invenio_config_tugraz/assets/semantic-ui/js/invenio_config_tugraz/notices.js
// This file is part of invenio-config-tugraz.
// Copyright (C) 2026 Graz University of Technology.
//
// invenio-config-tugraz is free software; you can redistribute it and/or
// modify it under the terms of the MIT License; see LICENSE file for more
// details.

import React, { useState } from "react";
import ReactDOM from "react-dom";
import { http } from "react-invenio-forms";

import "./notices.less";

const STORAGE_KEY = "tug-notices-acknowledged";

function acknowledgedLocally() {
  try {
    return JSON.parse(localStorage.getItem(STORAGE_KEY)) || [];
  } catch (e) {
    return [];
  }
}

function rememberLocally(key) {
  try {
    const keys = acknowledgedLocally();
    if (!keys.includes(key)) {
      localStorage.setItem(STORAGE_KEY, JSON.stringify([...keys, key]));
    }
  } catch (e) {
    // storage disabled
  }
}

function Notices({ notices, ackLabel, onAcknowledge }) {
  return (
    <>
      {notices.map((notice) => (
        <div
          key={notice.key}
          className={notice.accent ? "notice notice--accent" : "notice"}
        >
          <button
            type="button"
            className="notice-close"
            aria-label={ackLabel}
            onClick={() => onAcknowledge(notice.key)}
          >
            &times;
          </button>
          {notice.title && <h2>{notice.title}</h2>}
          {notice.intro && <p>{notice.intro}</p>}
          <ul>
            {notice.items.map((item, i) => (
              <li key={i}>{item}</li>
            ))}
          </ul>
          {notice.outro && <p className="notice-outro">{notice.outro}</p>}
          <button
            type="button"
            className="notice-dismiss"
            onClick={() => onAcknowledge(notice.key)}
          >
            {ackLabel}
          </button>
        </div>
      ))}
    </>
  );
}

function App({ initial, ackLabel, authenticated }) {
  const seen = authenticated ? [] : acknowledgedLocally();
  const [notices, setNotices] = useState(initial.filter((n) => !seen.includes(n.key)));

  const onAcknowledge = (key) => {
    setNotices((current) => current.filter((n) => n.key !== key));
    if (authenticated) {
      http.post(`/api/notices/${key}/acknowledge`);
    } else {
      rememberLocally(key);
    }
  };

  return <Notices notices={notices} ackLabel={ackLabel} onAcknowledge={onAcknowledge} />;
}

const el = document.getElementById("notices");
if (el) {
  ReactDOM.render(
    <App
      initial={JSON.parse(el.dataset.notices || "[]")}
      ackLabel={el.dataset.ackLabel}
      authenticated={el.dataset.authenticated === "true"}
    />,
    el
  );
}

File: invenio_config_tugraz/notices.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2026 Graz University of Technology.
#
# invenio-config-tugraz is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see LICENSE file for more
# details.

"""User notices, filtered by who they are shown to, role and per item."""

from flask import Blueprint, Flask, current_app, g, jsonify
from flask_login import current_user, login_required
from invenio_db import db
from werkzeug.wrappers import Response as BaseResponse

from .models import NoticeAcknowledgment


def current_roles() -> set[str]:
    """Role names held by the current identity."""
    return {need.value for need in g.identity.provides if need.method == "role"}


def is_shown_to(notice: dict, *, authenticated: bool) -> bool:
    """Check whether the notice should be shown to this visitor."""
    show_to = notice.get("show_to", "users")
    if show_to == "guests":
        return not authenticated
    if show_to == "users":
        return authenticated
    return True


def roles_match(roles: list | None, held: set[str]) -> bool:
    """Check whether the visitor holds one of the required roles."""
    return not roles or any(role in held for role in roles)


def visible_items(notice: dict, roles: set[str]) -> list[str]:
    """Return the item texts the visitor is allowed to see."""
    result = []
    for item in notice["items"]:
        if not isinstance(item, dict):
            result.append(str(item))
            continue
        if not roles_match(item.get("roles"), roles):
            continue
        check = item.get("visible")
        if check and not check():
            continue
        result.append(str(item["text"]))
    return result


def pending_notices() -> list[dict]:
    """Return the notices to show the current visitor."""
    authenticated = current_user.is_authenticated
    roles = current_roles() if authenticated else set()
    seen = set()
    if authenticated:
        seen = {
            ack.notice_key
            for ack in NoticeAcknowledgment.query.filter_by(user_id=current_user.id)
        }
    notices = []
    for notice in current_app.config.get("CONFIG_TUGRAZ_NOTICES", []):
        if not is_shown_to(notice, authenticated=authenticated):
            continue
        if authenticated and not roles_match(notice.get("roles"), roles):
            continue
        if notice["key"] in seen:
            continue
        notices.append(
            {
                "key": notice["key"],
                "title": str(notice["title"]),
                "intro": str(notice["intro"]),
                "items": visible_items(notice, roles),
                "outro": str(notice["outro"]) if notice.get("outro") else "",
                "accent": bool(notice.get("accent")),
            },
        )
    return notices


def api_blueprint(_app: Flask) -> Blueprint:
    """Blueprint for the notices API."""
    blueprint = Blueprint("tugraz_notices", __name__, url_prefix="/notices")

    @blueprint.route("/<key>/acknowledge", methods=["POST"])
    @login_required
    def acknowledge(key: str) -> BaseResponse:
        """Record that the current user acknowledged a notice."""
        already = NoticeAcknowledgment.query.filter_by(
            user_id=current_user.id,
            notice_key=key,
        ).count()
        if not already:
            db.session.add(
                NoticeAcknowledgment(user_id=current_user.id, notice_key=key),
            )
            db.session.commit()
        return jsonify({"acknowledged": key})

    return blueprint
